logo.js

```diff
@@ -818,40 +889,117 @@ function LogoInterpreter(turtle, stream, savehook)
     // Operate on a copy so the original is not destroyed
     statements = statements.slice();
 
-    var result;
-    while (statements.length) {
-      result = evaluateExpression(statements);
-
-      if (result !== undefined && !options.returnResult) {
-        throw new Error(format(__("Don't know what to do with {result}"), {result: result}));
+    return new Promise(function (resolve, reject) {
+      var lastResult;
+      function runLoop() {
+        if (self.forceBye) {
+          self.forceBye = false;
+          reject({special: "bye"});
+          return;
+        }
+        while (statements.length) {
+          var result = evaluateExpression(statements);
+          if (isPromise(result)) {
+            lastResult = result.then(function (value) {
+              lastResult = value;
+              runLoop();
+            }).catch(function (err) {
+              reject(err);
+            });
+            return;
+          }
+          lastResult = result;
+        }
+        if (lastResult !== undefined && !options.returnResult) {
+          reject(new Error(format(__("Don't know what to do with {result}"), {result: lastResult})));
+        }
+        resolve(lastResult);
       }
-    }
-
-    // Return last result
-    return result;
+      runLoop();
+    });
   };
 
+  // FIXME: should this confirm that something is running?
+  self.bye = function() {
+    self.forceBye = true;
+  };
 
   self.run = function(string, options) {
+    if (self._pendingPromise) {
+      var runner = self.run.bind(self, string, options);
+      return self._queueCall(runner);
+    }
     options = Object(options);
     if (self.turtle) { self.turtle.begin(); }
 
-    try {
-      // Parse it
-      var atoms = parse(string);
+    // Parse it
+    var atoms = parse(string);
```

Comment:
Does state get cleaned up if `parse()` fails, now that there's no try/catch around it?